### PR TITLE
Move to new release process and clean up Makefile

### PR DESCRIPTION
**Pull request by tomdee**
- Adjust some of the variable names to better reflect reality
- Stop pyinstaller from leaving unwanted files lying around (*.spec and
  build/)
- Build versioned and latest tagged images
- Simplify clean
- Use the upstream golang container to build calico/ctl
- Pass glide cache into the vendor target
- Delete a few files that shouldn't be checked in
- Document new release process

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,4 +1,4 @@
-.PHONY: all binary calico/node test ut ut-circle st st-ssl clean run-etcd run-etcd-ssl help clean_calico_node
```
Comment (robbrockbank) [inline Makefile]: I'd quite like to move the PHONY declarations to before the actual target - it's much easier to maintain.  Perhaps not one for this PR though.

File: RELEASING.md [hunk 2]
```diff
@@ -1,22 +1,24 @@
...
-Makefile: 
-- BUILD_CONTAINER_NAME?=calico/build:v0.18.0 - Currently, the startup.py script relies on the Python version of libcalico
-- FELIX_CONTAINER_NAME?=calico/felix:2.0.0-beta.3
-- LIBNETWORK_PLUGIN_CONTAINER_NAME?=calico/libnetwork-plugin:v1.0.0-beta-rc2
-- Also, less commonly Bird, GoBGP and confd
+# Release process
+
+## Resulting artifacts
+Creating a new release creates the following artifacts
+* `calico/node:$VERSION` container image
```
Comment (robbrockbank) [inline RELEASING.md]: +calico/node:latest
+calico/ctl:$VERSION

File: Makefile [hunk 3]
```diff
@@ -408,40 +407,61 @@ static-checks: vendor
...
-	-rm $(BUILD_CALICOCTL_CONTAINER_MARKER)
-	-rm $(CTL_CONTAINER_CREATED)
-	-docker rm -f calico-node
-	-docker rmi $(NODE_CONTAINER_NAME)
-	-docker rmi $(CTL_CONTAINER_NAME)
-	-docker rmi $(BUILD_CALICOCTL_CONTAINER_NAME)
-	-docker tag calico/test:latest calico/test:latest-backup && docker rmi calico/test:latest
-	-docker run -v /var/run/docker.sock:/var/run/docker.sock -v /var/lib/docker:/var/lib/docker --rm martin/docker-cleanup-volumes
-	-rm -r $(NODE_CONTAINER_DIR)/bin
+	rm -rf dist build certs *.tar vendor $(NODE_CONTAINER_DIR)/filesystem/bin
```
Comment (robbrockbank) [inline Makefile]: $NODE_CONTAINER_BIN

File: Makefile [hunk 4]
```diff
@@ -408,40 +407,61 @@ static-checks: vendor
...
+release: clean
+ifndef VERSION
+	$(error VERSION is undefined - run using make release VERSION=vX.Y.Z)
+endif
+	git tag $(VERSION)
+
+	$(MAKE) dist/calicoctl calico/ctl calico/node
+	# Check that the version output includes the version specified.
+# Tests that the "git tag" makes it into the binary. Main point is to catch "-dirty" builds
+	if ! dist/calicoctl version | grep -P 'Version:\s*$(VERSION)$$'; then echo "Reported version:" `dist/calicoctl version` "\nExpected version: $(VERSION)"; false; else echo "Version check passed\n"; fi
```
Comment (robbrockbank) [inline Makefile]: It would be nice to run this in the calico/ctl container - that way this should all just work in a OS X environment.

**Comment by robbrockbank** [comment]
CircleCI is also failing.

File: Makefile [hunk 5]
```diff
@@ -408,40 +407,61 @@ static-checks: vendor
...
+
+	# Check that images were created recently and that the IDs of the versioned and latest images match
+	@docker images --format "{{.CreatedAt}}\tID:{{.ID}}\t{{.Repository}}:{{.Tag}}" calico/node:latest
+	@docker images --format "{{.CreatedAt}}\tID:{{.ID}}\t{{.Repository}}:{{.Tag}}" calico/node:$(VERSION)
+	@docker images --format "{{.CreatedAt}}\tID:{{.ID}}\t{{.Repository}}:{{.Tag}}" calico/ctl:latest
+	@docker images --format "{{.CreatedAt}}\tID:{{.ID}}\t{{.Repository}}:{{.Tag}}" calico/ctl:$(VERSION)
 
-release-caliconode: calico/node
-	docker tag $(NODE_CONTAINER_NAME) quay.io/$(NODE_CONTAINER_NAME)
+	# Check that the images container the right sub-components
```
Comment (robbrockbank) [inline Makefile]: container -> contain

**Comment by robbrockbank** [comment]
make release was failing for me, running with VERSION=v2.0.0-beta.2  :  looks like the

# Check that the version output includes the version specified.
if ! dist/calicoctl version | grep -P 'Version:\s*v2.0.0-beta.2$'; then echo "Reported version:" `dist/calicoctl version` "\nExpected version: v2.0.0-beta.2"; false; else echo "Version check passed\n"; fi
Version:      v2.0.0-beta.2
Version check passed

# Check that images were created recently and that the IDs of the versioned and latest images match
2016-11-10 03:26:32 -0800 PST	ID:7c5a3743c34a	calico/node:latest
2016-11-10 03:26:32 -0800 PST	ID:7c5a3743c34a	calico/node:v2.0.0-beta.2
2016-11-11 04:29:57 -0800 PST	ID:91b9b1e36f82	calico/ctl:latest
2016-11-11 04:29:57 -0800 PST	ID:91b9b1e36f82	calico/ctl:v2.0.0-beta.2
# Check that the images container the right sub-components
docker run calico/node:v1.0.0-beta-rc4-30-g975b01b calico-felix --version
Unable to find image 'calico/node:v1.0.0-beta-rc4-30-g975b01b' locally
Pulling repository docker.io/calico/node
docker: Tag v1.0.0-beta-rc4-30-g975b01b not found in repository docker.io/calico/node.
See 'docker run --help'.
Makefile:414: recipe for target 'release' failed
make: *** [release] Error 125

File: RELEASING.md [hunk 6]
```diff
@@ -1,22 +1,24 @@
...
+- `FELIX_CONTAINER_NAME?=calico/felix:2.0.0`
+- `LIBNETWORK_PLUGIN_CONTAINER_NAME?=calico/libnetwork-plugin:v1.0.0`
+- Also, less commonly Bird, GoBGP and confd
 
-After performing these updates, make sure they are tested and commit the changes to `master`
+You should have no local changes and tests should be passing.
 
-## Releasing calico-containers (`calico/node` and `calicoctl`
+## Creating the release
 1. Choose a version e.g. `export VERSION=v1.0.0`
```
Comment (robbrockbank) [inline RELEASING.md]: you don't need to export VERSION do you?

Comment (robbrockbank) [inline RELEASING.md]: Do we need to have an environment variable set *and* pass it in on the make target?  I'd rather we just forced the user to specify it on the make target and got rid of the environment variable if possible.

**Comment by djosborne** [comment]
@tomdee Wanted to bring #1288 to your attention. Does this PR address it?

**Comment by tomdee** [comment]
@djosborne It does not directly, but it might indirectly

**Pull request merged**